### PR TITLE
Enforce removeSegment flow with _enableDeletedKeysCompactionConsistency

### PR DESCRIPTION
**Pull request by tibrewalpratik17**
When we enable `_enableDeletedKeysCompactionConsistency`, we want the `removeSegment` process to run to decrease the `distinctSegmentCount` by 1. Previously, in issue #13347, it was assumed that during the `replaceSegment` operation (whether during commit or refresh segment flow), the `addSegment` method would increase the count by 1, and `removeSegment` would decrease it by 1. However, it’s not guaranteed that the `removeSegment` process will be triggered during the `replaceSegment` operation.

Here are the steps currently followed during replaceSegment in the `ConcurrentMapPartitionUpsertMetadataManager` class:
- Add the new segment.
- Check if any keys remain unreplaced in the old segment. If any are found, delete those keys from the hashmap using `removeSegment(oldSegment, remainingValidDocIDsSnapshotNotReplaced)` [[Ref](https://github.com/apache/pinot/blob/e9b898d7ed59610afcb64b2ea30153017e9aa9e0/pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/BasePartitionUpsertMetadataManager.java#L722-L736)].
- Remove oldSegment from _trackedSegments [[Ref](https://github.com/apache/pinot/blob/e9b898d7ed59610afcb64b2ea30153017e9aa9e0/pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/BasePartitionUpsertMetadataManager.java#L643)].

During the destroy process of SegmentDataManager, the removeSegment(segment) method is called. However, this becomes a no-op since the segment has already been removed from _trackedSegments in step 3 [[Ref](https://github.com/apache/pinot/blob/e9b898d7ed59610afcb64b2ea30153017e9aa9e0/pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/BasePartitionUpsertMetadataManager.java#L760-L764)].

For `_enableDeletedKeysCompactionConsistency`, it’s crucial to ensure that `removeSegment(segment)` is called to maintain consistency in the distinctSegmentCount value. This change overrides the replaceSegment method in the `ConcurrentMapPartitionUpsertMetadataManagerForConsistentDeletes` class, ensuring that the removeSegment(segment) process is always executed at the end.

Additionally, I updated the UTs to track the value of distinctSegmentCount, which was missed in #13347.

After deploying this change, I successfully tested that the `deletedttlkeysinmultiplesegments` metric decreased significantly, as shown in the screenshot below, indicating that the keys are now being deleted properly and consistently. Previously, the distinctSegmentCount was only increasing by 1 without decreasing, leading to the keys not being deleted as expected.

Screenshot 2024-08-30 at 12 29 16 AM

### CODE REVIEW

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/BasePartitionUpsertMetadataManager.java [hunk 1]
```diff
@@ -544,9 +544,7 @@ public void addSegment(ImmutableSegmentImpl segment, @Nullable ThreadSafeMutable
   protected void addOrReplaceSegment(ImmutableSegmentImpl segment, ThreadSafeMutableRoaringBitmap validDocIds,
       @Nullable ThreadSafeMutableRoaringBitmap queryableDocIds, Iterator<RecordInfo> recordInfoIterator,
       @Nullable IndexSegment oldSegment, @Nullable MutableRoaringBitmap validDocIdsForOldSegment) {
-    if (_partialUpsertHandler != null) {
```
Comment (klsince) [inline pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/BasePartitionUpsertMetadataManager.java:547]: why this change?

Comment (tibrewalpratik17) [reply pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/BasePartitionUpsertMetadataManager.java:547]: Keeping the behaviour same for partial-upsert and full-upsert tables. This is not causing any behavioural changes for full-upsert tables.

Comment (klsince) [reply pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/BasePartitionUpsertMetadataManager.java:547]: hmm. but will this be extra cost for full upsert table? as the resolveComparisonTies method iterates records

Comment (tibrewalpratik17) [reply pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/BasePartitionUpsertMetadataManager.java:547]: It should be compensated with the update record location part in ConcurrentHashmap flow as say instead of x record-location reference updates per key we will only have 1.

Comment (tibrewalpratik17) [reply pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/BasePartitionUpsertMetadataManager.java:547]: Yeah true! Let me know if you want to revert this. 
I was planning to keep the behaviour same for partial-upsert and full-upsert though.

Comment (klsince) [reply pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/BasePartitionUpsertMetadataManager.java:547]: sry for the delay. yes, I'd prefer to keep the if-check to resolve ties for partial upsert.

Comment (tibrewalpratik17) [reply pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/BasePartitionUpsertMetadataManager.java]: Done! I had to move this inside the sub-class as i realised why I added this! For full-upsert, if we don't call `resolveComparisonTies` then we might end-up incrementing the segment-count multiple times but decreasing it only once during `replaceSegment` flow.

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/ConcurrentMapPartitionUpsertMetadataManagerForConsistentDeletes.java [hunk 2]
```diff
@@ -81,6 +85,12 @@ protected long getNumPrimaryKeys() {
   protected void doAddOrReplaceSegment(ImmutableSegmentImpl segment, ThreadSafeMutableRoaringBitmap validDocIds,
       @Nullable ThreadSafeMutableRoaringBitmap queryableDocIds, Iterator<RecordInfo> recordInfoIterator,
       @Nullable IndexSegment oldSegment, @Nullable MutableRoaringBitmap validDocIdsForOldSegment) {
+    if (_partialUpsertHandler == null) {
+      // for full upsert, we are de-duping primary key once here to make sure that we are not adding
+      // primary-key multiple times and subtracting just once in removeSegment.
+      // for partial-upsert, we call this method in base class.
+      recordInfoIterator = resolveComparisonTies(recordInfoIterator, _hashFunction);
```
Comment (klsince) [inline pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/ConcurrentMapPartitionUpsertMetadataManagerForConsistentDeletes.java:92]: 👍 make sense

**Review by klsince — APPROVED** [review]
